hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerDoubleBuffer.java

```diff
@@ -572,12 +572,6 @@ private synchronized boolean canFlush() {
       return true;
     } catch (InterruptedException ex) {
       Thread.currentThread().interrupt();
-      if (isRunning.get()) {
```

Comment:
In [stopDaemon()](https://github.com/apache/ozone/blob/87d8d6155ca1d68773061017385c11fcfde43214/hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerDoubleBuffer.java#L516), we first set `isRunning` to false and then interrupt the thread. So thread cannot be interrupted while it is running. And if that's happening, it should be fine to terminate the service.

Reply:
There is a chance this can happen,
1. canFlush() keeps waiting for events continuously
2. stopDaemon set isRunning to false and trigger interrupt
3. this can interrupt canFlush on wait()


```
while (currentBuffer.isEmpty()) {
        // canFlush() only gets called when the readyBuffer is empty.
        // Since both buffers are empty, notify once for each.
        flushNotifier.notifyFlush();
        flushNotifier.notifyFlush();
        wait(1000L);
      }
```

Reply:
I have done more changes to handle missing of interrupt signal case.
if interrupt is triggered but thread is not on wait, it can miss the interrupt signal. So additional isRunning check is added to ensure its exit.

Reply:
> canFlush() keeps waiting for events continuously
> ...
> this can interrupt canFlush on wait()

Isn't that exactly the purpose of the interrupt?

`ExitUtils.terminate` is called here only if two conditions are both met:

- the thread is interrupted
- `isRunning == true`

Since `stopDaemon()` interrupts only after setting `isRunning = false`, there are two possibilities:

1. thread is interrupted by some other code
2. `isRunning = true` is set by some other code after `stopDaemon()` set it to `false`, but before handling the `InterruptedException`

@sumitagrawl can you share logs where this happened?

Reply:
@adoroszlai 

From code review perspective, pause() should not trigger Exit of OM Service. The pause() is triggered when running installSnapshot for the OM node. Same is planned to use for other feature.

Interrupt() purpose is just only to exit the thread. But should not be termination of OM, unless it is causing some inconsistency. So logic seems incorrect where service is getting terminated on interrupt as no inconsistency in this case. Its job is graceful stopping.

`stopDaemon()` itself is triggering interrupt to exit the thread immediately after setting `isRunning = false`.

```
 @VisibleForTesting
  public void stopDaemon() {
    if (isRunning.compareAndSet(true, false)) {
      LOG.info("Stopping OMDoubleBuffer flush thread");
 ==>     daemon.interrupt();    ///,--- this is here interrupt triggered
      try {
        // Wait for daemon thread to exit
        daemon.join();
      } catch (InterruptedException e) {
```

As current code, no one reset `isRunning = true` again. So there is no such scenario as mentioned above.

Reply:
> `stopDaemon()` itself is triggering interrupt to exit the thread immediately after setting `isRunning = false`.
> 
> ```
>  @VisibleForTesting
>   public void stopDaemon() {
>     if (isRunning.compareAndSet(true, false)) {
>       LOG.info("Stopping OMDoubleBuffer flush thread");
>  ==>     daemon.interrupt();    ///,--- this is here interrupt triggered
>       try {
>         // Wait for daemon thread to exit
>         daemon.join();
>       } catch (InterruptedException e) {
> ```

This is the case where OM is not terminated, since `if (isRunning.get())` block is not entered.  Please read my comment again.

Reply:
> 1. thread is interrupted by some other 
Currently no one is giving an interrupt, checked the code. So no such case.

> 2. isRunning = true is set by some other code after stopDaemon() set it to false, but before handling the InterruptedException
This also, there is no case present in code doing same

So unable to find any such case.

Reply:
> unable to find any such case

Your code analysis shows (and I find the same) that the call to `terminate` cannot happen.  If you have encountered such case, two things are possible:

1. our analysis is wrong (and the bug is still there)
2. the code that resulted in OM termination is different than the one we analyzed (and the bug is already fixed)

In either case, removing `terminate()` is not fixing the root cause.  It will only make the problem harder to detect.

Reply:
Since terminate is called only when isRunning is true, which is not possible. So this PR is not required.